### ISSUE TITLE
Does not work on windows

### ISSUE THREAD

**Issue by JSH32** [coder/code-server#2388]
When trying to access code-server through the proxy It is unable to load any of the extensions and checking the log shows the results below.
![Screenshot](https://i.kawaii.sh/iChrFpO.png)
This does not happen when I try to access it from my ubuntu laptop, it only fails on Windows.

- Web Browser: Chrome
- Local OS: Windows
- Remote OS: Ubuntu 20.04
- Remote Architecture: x64
- `code-server --version`: 3.7.3 93fb76e4a71b1959ec2a23481ba2611200d714b2

**Comment by Riku32** [comment]
I have tested this on all browsers including Firefox. The same issue occurs only specifically on windows.

**Comment by Riku32** [comment]
I have just updated to 3.7.4, the same bug occurs, I am somewhat inclined to believe this is because of windows firewall but it uses standard ports so I have become more confused. Looking into the issue I am seeing a few new errors pop up.
![Screenshot](https://i.kawaii.sh/7QXd_aU.png)
On ubuntu the sourcemap is also not found but the ExtensionHost loads and shows the entire output of plugins being initialized. It seems to be failing on this block of code
```ts
this._proxy = this._extensionHost.start()!.then(
			(protocol) => {
				this._hasStarted = true;
				return { value: this._createExtensionHostCustomers(protocol) };
			},
			(err) => {
				console.error('Error received from starting extension host');
				console.error(err);
				return null;
			}
		);
```
`_extensionHost` appears to be implementing interface `IExtensionHost`. I am guessing that the method `.start()` of this interface is was originally platform specific VSCode code that code-server is yet to replace or has accidentally allowed to use. This can probably be assumed to happen due to platform specific code in VS.
Hopefully this provides some kind of help on how this issue can be fixed.

**Comment by code-asher** [comment]
The source map is a bug; I'll be fixing that soon but it shouldn't cause any issues.

I tried using Chrome+Windows accessing 3.7.3 running on a Linux host but everything connected and I'm able to use it normally (all done locally).

Could you run with `--log debug` then post the browser logs and code-server's stdout?

**Comment by Riku32** [comment]
After a few times of it failing to login and retrying the logs are
```
[2020-12-04T00:06:17.802Z] debug forking vs code...
[2020-12-04T00:06:18.095Z] debug setting up vs code...
debug vscode got message from code-server {"type":"init"}
debug vscode got message from code-server {"type":"socket"}
debug vscode New connection {"token":"this is something"}
[2020-12-04T00:06:22.735Z] debug got latest version {"latest":"3.7.4"}
[2020-12-04T00:06:22.737Z] debug comparing versions {"current":"3.7.4","latest":"3.7.4"}
debug vscode got message from code-server {"type":"socket"}
debug vscode New connection {"token":"this is different than the last"}
debug vscode got message from code-server {"type":"socket"}
debug vscode got message from code-server {"type":"socket"}
debug vscode got message from code-server {"type":"socket"}
debug vscode got message from code-server {"type":"socket"}
```

**Comment by Riku32** [comment]
I have also tested not using the coder proxy. It loads perfectly when directly connecting.

**Comment by code-asher** [comment]
Hmm, in that case it might be the proxy that is the issue if connecting
directly works. What proxy are you using? If you post the config we can
take a look and see if anything looks amiss.

Although you mentioned it worked from another laptop; is that one also
going through the proxy?

**Comment by code-asher** [comment]
Oh I misread, you're using the coder proxy with `--link`?

**Comment by Riku32** [comment]
I used `--link` yes

**Comment by code-asher** [comment]
I'm not entirely sure what's going on in that case since I haven't been
able to replicate yet. Would you be able to post the full browser log?
It might also be helpful to look at the network tab and see if any of
the requests are failing.

**Comment by L3P3** [comment]
My server is debian bullseye, I just upgraded to latest code-server version, from the version I installed a month ago. My client is Windows XP, latest Firefox. When I open the page, I get an error message dialog telling me that it was unable to connect and I need to reload the page. Does not help thou. No idea if this issue is related.

**Comment by nhooyr** [comment]
@L3P3 Seems related feel free to drop more info in this issue.

**Comment by L3P3** [comment]
My issue is independent of Windows. When I open the page on debian, I get the same issue. One thing I noticed:
I upgraded (installed the new .deb file over the previous version) but when I open the page, it still shows me the previous version there. I tried to stop the service, killall code-server but still, I can open the page and it shows the old version. I am rebooting my server now.

**Comment by nhooyr** [comment]
Hmm, try resetting your cache as well!

**Comment by L3P3** [comment]
It worked by restarting my server. The old server process was not properly killed when I installed the new package over it.

**Comment by L3P3** [comment]
Nope. It shows up the new version but still disconnects, no matter if my client is Windows or not.

**Comment by nhooyr** [comment]
Nooo!! I was excited.

**Comment by L3P3** [comment]
Now it is working again... I stopped the systemd service and instead ran code-server in terminal. It works now. But I will have to see how long.

**Comment by L3P3** [comment]
It also works when I open the page from Firefox running in Windows. Maybe I misunderstood the issue here.

**Comment by Riku32** [comment]
How would I submit my browser log, it is still happening even though I have reinstalled windows. Exactly the screenshot above.

**Comment by code-asher** [comment]
I think the easiest way is to right click the console and use the "save
as" option then attach the saved file here.

**Comment by Riku32** [comment]
New information, I have tried it on a computer that is not mine with the same install of windows and the issue was resolved.

**Comment by code-asher** [comment]
Definitely strange! Was that other computer you tried on the same
network? Maybe it's network related somehow.

**Comment by Riku32** [comment]
It was on a different network, I originally suspected windows firewall but it was the same version of windows, theirs was not a fresh install though. It works on other operating systems on my network.

**Comment by shayne** [comment]
I've been able to reproduce this issue. I have code-server running in the cloud, accessed via private IP over a WireGuard connection.

My local subnet is 10.2.0.0/16 and my remote private subnet is 10.3.0.0/16. Hosting cloud-server within the local 10.2.X.X subnet works fine, accessing via 10.3.X.X with default Windows Firewall (connection set to Private) results in the reconnect error.

I can reproduce the issue by turning off the Windows Firewall and loading code-server. The connection comes up right away and extensions start running. Note: after this initial connection code-server will continue to work for some time after re-enabling the firewall. I believe this is due to connection state tracking.

I turned on logging and didn't see any related DROP messages. I'm continuing to look into the issue. Ideally I'd like to find a solution that doesn't involve modifying the firewall for each client.

Ideas?

**Comment by L3P3** [comment]
Yes, the Windows Firewall as most is stateful so it memorizes outgoing connections for a while to let replies in. Maybe there is a message from the server to the client after a too long delay. Maybe more frequent pings from the client to the server could fix the issue. But I am just roughly guessing, I am not an expert in sockets and firewalls at all. :wink:

**Comment by jsjoeio** [comment]
I haven't used Windows in a long time so not the expert here but maybe @code-asher has some ideas

**Comment by shayne** [comment]
I've got it working without modifying the Windows Firewall.

I'm not sure the exact issue, but I'll share my findings and the fix here in hopes it sheds light or helps someone else.

I ran Wireshark to see what was going on when loading code-server in the browser. One thing that stood out to me were the repeated ICMP Redirects. These caused by having a static route directly traffic to my remote subnet 10.3.0.0/16.

To avoid the redirect I added a static route to my machine that bypassed my router. To my surprise code-server made connection and extensions fired up. It worked.

So two things: 1) it appears as if Windows Firewall ignores the redirects by default and doesn't modify the local route table; 2) code-server wasn't liking the redirect or going through the router for the extension host connection.

I modified my DHCP server to hand out static routes using the DHCP option "121" and providing the hex encoded string of routes. It's working fine now.

I'll also note that I use many other services (web, ssh, routing, etc...) on the 10.3.X.X network and haven't experienced any issues up until this code-server bug.
